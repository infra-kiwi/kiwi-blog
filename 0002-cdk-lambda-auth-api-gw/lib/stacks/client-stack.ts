import * as cdk from 'aws-cdk-lib';
import { aws_iam, aws_lambda, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as path from 'node:path';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Effect } from 'aws-cdk-lib/aws-iam';

export type ClientStackProps = cdk.StackProps & {
  serverGatewayUrl: string;
  serverGatewayRegion: string;
};

/**
 * This Stack defines the API Gateway and the Client Lambda
 */
export class ClientStack extends cdk.Stack {
  readonly props: ClientStackProps;

  constructor(scope: Construct, id: string, props: ClientStackProps) {
    super(scope, id, props);
    this.props = props;

    // Define the Client Lambda function
    const lambda = new NodejsFunction(this, 'ClientLambda', {
      functionName: `${id}-client-lambda`,
      entry: path.join(__dirname, '..', 'lambda', 'client.ts'),
      runtime: aws_lambda.Runtime.NODEJS_20_X,
      bundling: {
        externalModules: ['@aws-sdk/*'],
        minify: false
      },
      // The Lambda entrypoint name
      handler: 'index.handler',

      // Pass the server gateway info as env vars
      environment: {
        SERVER_GATEWAY_URL: props.serverGatewayUrl,
        SERVER_GATEWAY_REGION: props.serverGatewayRegion
      }
    });
    // Allow the lambda to make API Gateway calls in general
    // NOTE: This should be scoped more properly for any production deployment!
    lambda.addToRolePolicy(
      new aws_iam.PolicyStatement({
        effect: Effect.ALLOW,
        actions: ['execute-api:Invoke'],
        resources: ['*']
      })
    );

    // Record the name for ease of usage of the example
    new CfnOutput(this, 'ClientLambdaFunctionName', {
      value: lambda.functionName
    });
  }
}
