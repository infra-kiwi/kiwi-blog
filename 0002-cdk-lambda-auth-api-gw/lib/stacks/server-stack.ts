import * as cdk from 'aws-cdk-lib';
import { Aws, aws_iam, aws_lambda, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as path from 'node:path';
import {
  AuthorizationType,
  EndpointType,
  LambdaIntegration,
  RestApi
} from 'aws-cdk-lib/aws-apigateway';
import {
  AccountPrincipal,
  PolicyDocument,
  PolicyStatement
} from 'aws-cdk-lib/aws-iam';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';

export type ServerStackProps = cdk.StackProps & {
  // The AWS account id of the client account
  allowedPrincipals: aws_iam.IPrincipal[];
};

/**
 * This Stack defines the API Gateway and the Server Lambda
 */
export class ServerStack extends cdk.Stack {
  readonly id: string;
  readonly props: ServerStackProps;
  readonly apiGatewayPolicy: PolicyDocument;
  readonly apiGateway: RestApi;

  constructor(scope: Construct, id: string, props: ServerStackProps) {
    super(scope, id, props);
    this.id = id;
    this.props = props;

    // Define the IAM policy that will allow clients to send
    // requests to this API Gateway via IAM
    this.apiGatewayPolicy = new PolicyDocument({
      statements: [
        new PolicyStatement({
          actions: ['execute-api:Invoke'],
          // By default, allow all invocations from inside the Server account
          principals: [new AccountPrincipal(Aws.ACCOUNT_ID)]
        })
      ]
    });

    // Define the API Gateway resource
    this.apiGateway = new RestApi(this, 'Gateway', {
      endpointTypes: [EndpointType.REGIONAL],
      policy: this.apiGatewayPolicy
    });

    // Export the API Gateway url and region, which we can then
    // use in the client function
    new CfnOutput(this, 'ServerGatewayUrl', {
      value: this.apiGateway.url
    });
    new CfnOutput(this, 'ServerGatewayRegion', {
      value: Aws.REGION
    });

    // Define all the routes
    this.#renderLambdaRoute();
  }

  #renderLambdaRoute() {
    // Define the Server Lambda function
    const lambda = new NodejsFunction(this, 'ServerLambda', {
      functionName: `${this.id}-server-lambda`,
      entry: path.join(__dirname, '..', 'lambda', 'server.ts'),
      runtime: aws_lambda.Runtime.NODEJS_20_X,
      // The Lambda entrypoint name
      handler: 'index.handler',
      bundling: {
        externalModules: ['@aws-sdk/*'],
        minify: false
      }
    });

    const routePath = 'my-lambda';

    // Add the Server Lambda function as a target in the API Gateway, exposed under
    // the `/my-lambda` path with a GET method
    const resource = this.apiGateway.root.addResource(routePath);
    resource.addMethod('POST', new LambdaIntegration(lambda), {
      // Require the HTTP calls to be IAM-authorized
      authorizationType: AuthorizationType.IAM
    });

    // Compose the ARN of the Server Lambda route
    // See https://docs.aws.amazon.com/apigateway/latest/developerguide/arn-format-reference.html
    const routeArn = `arn:aws:execute-api:${Aws.REGION}:${Aws.ACCOUNT_ID}:*/*/POST/${routePath}`;
    // Alternatively, if you want to allow access to ALL API Gateway routes
    // const routeArn = `arn:aws:execute-api:${Aws.REGION}:${Aws.ACCOUNT_ID}:*`;

    // Allow the access to this route from the desired principals
    // (e.g., the client account or organization)
    this.apiGatewayPolicy.addStatements(
      new PolicyStatement({
        principals: this.props.allowedPrincipals,
        actions: ['execute-api:Invoke'],
        resources: [routeArn]
      })
    );
  }
}
