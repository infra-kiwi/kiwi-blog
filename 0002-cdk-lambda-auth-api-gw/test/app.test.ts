import { App, Aws } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { ServerStack } from '../lib/stacks/server-stack';
import { OrganizationPrincipal } from 'aws-cdk-lib/aws-iam';
import { ClientStack } from '../lib/stacks/client-stack';

test('Stacks', async () => {
  const orgId = 'o-12345abcde';

  const app = new App();

  // --- Server ---
  const serverStack = new ServerStack(app, 'Server', {
    allowedPrincipals: [new OrganizationPrincipal(orgId)]
  });

  // --- Client ---

  const clientStack = new ClientStack(app, 'Client', {
    serverGatewayUrl: serverStack.apiGateway.url,
    serverGatewayRegion: Aws.REGION
  });

  // --- Tests ---

  const serverTemplate = Template.fromStack(serverStack);
  expect(serverTemplate).toMatchSnapshot();
  serverTemplate.resourceCountIs('AWS::ApiGateway::RestApi', 1);

  const clientTemplate = Template.fromStack(clientStack);
  expect(clientTemplate).toMatchSnapshot();
});
