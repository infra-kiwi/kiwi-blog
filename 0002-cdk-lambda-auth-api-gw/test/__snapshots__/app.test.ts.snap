// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stacks 1`] = `
{
  "Outputs": {
    "ExportsOutputRefGateway43CB3636182B6A60": {
      "Export": {
        "Name": "Server:ExportsOutputRefGateway43CB3636182B6A60",
      },
      "Value": {
        "Ref": "Gateway43CB3636",
      },
    },
    "ExportsOutputRefGatewayDeploymentStageprod62BC69044D49F930": {
      "Export": {
        "Name": "Server:ExportsOutputRefGatewayDeploymentStageprod62BC69044D49F930",
      },
      "Value": {
        "Ref": "GatewayDeploymentStageprod62BC6904",
      },
    },
    "GatewayEndpoint4DF49EE0": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "Gateway43CB3636",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "GatewayDeploymentStageprod62BC6904",
            },
            "/",
          ],
        ],
      },
    },
    "ServerGatewayRegion": {
      "Value": {
        "Ref": "AWS::Region",
      },
    },
    "ServerGatewayUrl": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "Gateway43CB3636",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "GatewayDeploymentStageprod62BC6904",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Gateway43CB3636": {
      "Properties": {
        "EndpointConfiguration": {
          "Types": [
            "REGIONAL",
          ],
        },
        "Name": "Gateway",
        "Policy": {
          "Statement": [
            {
              "Action": "execute-api:Invoke",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
            {
              "Action": "execute-api:Invoke",
              "Condition": {
                "StringEquals": {
                  "aws:PrincipalOrgID": "o-12345abcde",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:execute-api:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":*/*/POST/my-lambda",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "GatewayAccount6D7AA49C": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "Gateway43CB3636",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "GatewayCloudWatchRole34305E06",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "GatewayCloudWatchRole34305E06": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "GatewayDeployment880EB8C2ff1a3b73546e97e19dfcd860276d2feb": {
      "DependsOn": [
        "GatewaymylambdaPOST74714818",
        "GatewaymylambdaC0E83592",
      ],
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "Gateway43CB3636",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "GatewayDeploymentStageprod62BC6904": {
      "DependsOn": [
        "GatewayAccount6D7AA49C",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "GatewayDeployment880EB8C2ff1a3b73546e97e19dfcd860276d2feb",
        },
        "RestApiId": {
          "Ref": "Gateway43CB3636",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "GatewaymylambdaC0E83592": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "Gateway43CB3636",
            "RootResourceId",
          ],
        },
        "PathPart": "my-lambda",
        "RestApiId": {
          "Ref": "Gateway43CB3636",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "GatewaymylambdaPOST74714818": {
      "Properties": {
        "AuthorizationType": "AWS_IAM",
        "HttpMethod": "POST",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "ServerLambdaECD01B55",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "GatewaymylambdaC0E83592",
        },
        "RestApiId": {
          "Ref": "Gateway43CB3636",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "GatewaymylambdaPOSTApiPermissionServerGateway416220A0POSTmylambdaB82AF6BB": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ServerLambdaECD01B55",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "Gateway43CB3636",
              },
              "/",
              {
                "Ref": "GatewayDeploymentStageprod62BC6904",
              },
              "/POST/my-lambda",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "GatewaymylambdaPOSTApiPermissionTestServerGateway416220A0POSTmylambda5244080A": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ServerLambdaECD01B55",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "Gateway43CB3636",
              },
              "/test-invoke-stage/POST/my-lambda",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ServerLambdaECD01B55": {
      "DependsOn": [
        "ServerLambdaServiceRoleC84284B4",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "e7d2282f762836d0ac164f863e9d87b335a12369e8cf280816e2f7b859b75180.zip",
        },
        "FunctionName": "Server-server-lambda",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "ServerLambdaServiceRoleC84284B4",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ServerLambdaServiceRoleC84284B4": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Stacks 2`] = `
{
  "Outputs": {
    "ClientLambdaFunctionName": {
      "Value": {
        "Ref": "ClientLambda1FF668DE",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ClientLambda1FF668DE": {
      "DependsOn": [
        "ClientLambdaServiceRoleDefaultPolicyCFDB40F4",
        "ClientLambdaServiceRoleF6F3BAE1",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "37a710945d7da5c4037e1dc864eb348196057b163c3e2c4f0a6e2a41b81a87f9.zip",
        },
        "Environment": {
          "Variables": {
            "SERVER_GATEWAY_REGION": {
              "Ref": "AWS::Region",
            },
            "SERVER_GATEWAY_URL": {
              "Fn::Join": [
                "",
                [
                  "https://",
                  {
                    "Fn::ImportValue": "Server:ExportsOutputRefGateway43CB3636182B6A60",
                  },
                  ".execute-api.",
                  {
                    "Ref": "AWS::Region",
                  },
                  ".",
                  {
                    "Ref": "AWS::URLSuffix",
                  },
                  "/",
                  {
                    "Fn::ImportValue": "Server:ExportsOutputRefGatewayDeploymentStageprod62BC69044D49F930",
                  },
                  "/",
                ],
              ],
            },
          },
        },
        "FunctionName": "Client-client-lambda",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "ClientLambdaServiceRoleF6F3BAE1",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ClientLambdaServiceRoleDefaultPolicyCFDB40F4": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "execute-api:Invoke",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ClientLambdaServiceRoleDefaultPolicyCFDB40F4",
        "Roles": [
          {
            "Ref": "ClientLambdaServiceRoleF6F3BAE1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ClientLambdaServiceRoleF6F3BAE1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
